imports/server/PublicationMerger.ts

```diff
@@ -0,0 +1,272 @@
+      if (existing === values[0] && !EJSON.equals(existing.value, value)) {
+        accumulator[key] = value;
+      }
+      existing.value = value;
+    } else {
+      values.push({ handle, value });
+    }
+  }
+}
+
+class CollectionView {
+  name: string;
+
+  callbacks: Pick<Subscription, 'added' | 'changed' | 'removed'>;
+
+  documents: Map<string /* id */, DocumentView> = new Map();
+
+  constructor(name: string, callbacks: Pick<Subscription, 'added' | 'changed' | 'removed'>) {
+    this.name = name;
+    this.callbacks = callbacks;
+  }
+
+  removeSubscription(handle: Subscription) {
+    for (const [id, document] of this.documents.entries()) {
+      if (document.handles.has(handle)) {
+        this.removed(handle, id);
+      }
+    }
+  }
+
+  added(handle: Subscription, id: string, fields: Record<string, any>) {
+    let document = this.documents.get(id);
+    let added = false;
+    if (!document) {
+      added = true;
+      document = new DocumentView();
+      this.documents.set(id, document);
+    }
+
+    document.handles.add(handle);
+
+    const changes: Record<string, any> = {};
+    for (const [key, value] of Object.entries(fields)) {
+      document.changeField(handle, key, value, changes, true);
+    }
+    if (added) {
+      this.callbacks.added(this.name, id, changes);
+    } else {
+      this.callbacks.changed(this.name, id, changes);
+    }
+  }
+
+  changed(handle: Subscription, id: string, fields: Record<string, any>) {
+    const changes: Record<string, any> = {};
+    const document = this.documents.get(id);
+    if (!document) {
+      throw new Error('changed called on unpublished document');
+    }
+    for (const [key, value] of Object.entries(fields)) {
+      if (value === undefined) {
+        document.clearField(handle, key, changes);
+      } else {
+        document.changeField(handle, key, value, changes);
+      }
+    }
+    this.callbacks.changed(this.name, id, changes);
+  }
+
+  removed(handle: Subscription, id: string) {
+    const document = this.documents.get(id);
+    if (!document) {
+      throw new Error('removed called on unpublished document');
+    }
+    document.handles.delete(handle);
+    if (document.handles.size === 0) {
+      this.documents.delete(id);
+      this.callbacks.removed(this.name, id);
+    } else {
+      const changes: Record<string, any> = {};
+      for (const key of document.data.keys()) {
+        document.clearField(handle, key, changes);
+      }
+      this.callbacks.changed(this.name, id, changes);
+    }
+  }
+}
+
+class SubSubscription implements Subscription {
+  merger: PublicationMerger;
+
+  connection: Meteor.Connection;
+
+  userId: string | null;
+
+  stopCallbacks: (() => void)[] = [];
+
+  constructor(merger: PublicationMerger) {
+    this.merger = merger;
+    this.connection = merger.mainSub.connection;
+    this.userId = merger.mainSub.userId;
+  }
+
+  ready!: () => void;
+
+  readyPromise: Promise<void> = new Promise<void>((resolve) => {
```

Comment:
This appears to only work with the most-recently-created `readyPromise`.  If that's what you want, maybe document that behavior, but I'd find it more well-behaved if `ready()` armed *all* the `readyPromise`s that have been created so far.

Also: if I'm reading this right, this appears to only work if you create the Promise before `ready()` gets called?  I'd have expected a later call to `readyPromise` to get a promise that is immediately resolved if the sub is ready already, rather than one that hangs potentially indefinitely.

Reply:
This is an initializer, so the promise gets created once when the class is instantiated.

Reply:
Oh.  Right, sorry for the noise 😅